ibis/backends/dask/trace.py

```diff
@@ -0,0 +1,147 @@
+DEBUG:ibis.dask.trace:       execute_database_table_client DaskTable
+DEBUG:ibis.dask.trace:       execute_database_table_client DaskTable 0:00:00.000085
+DEBUG:ibis.dask.trace:     execute_until_in_scope DaskTable 0:00:00.000362
+DEBUG:ibis.dask.trace:     execute_selection_dataframe Selection
+DEBUG:ibis.dask.trace:       main_execute ElementWiseVectorizedUDF
+DEBUG:ibis.dask.trace:         execute_until_in_scope ElementWiseVectorizedUDF
+DEBUG:ibis.dask.trace:           execute_until_in_scope TableColumn
+DEBUG:ibis.dask.trace:             execute_until_in_scope DaskTable
+DEBUG:ibis.dask.trace:             execute_until_in_scope DaskTable 0:00:00.000061
+DEBUG:ibis.dask.trace:             execute_table_column_df_or_df_groupby TableColumn
+DEBUG:ibis.dask.trace:             execute_table_column_df_or_df_groupby TableColumn 0:00:00.000304  # noqa: E501
+DEBUG:ibis.dask.trace:           execute_until_in_scope TableColumn 0:00:00.000584
+DEBUG:ibis.dask.trace:           execute_udf_node ElementWiseVectorizedUDF
+DEBUG:ibis.dask.trace:           execute_udf_node ElementWiseVectorizedUDF 0:00:05.019173
+DEBUG:ibis.dask.trace:         execute_until_in_scope ElementWiseVectorizedUDF 0:00:05.052604  # noqa: E501
+DEBUG:ibis.dask.trace:       main_execute ElementWiseVectorizedUDF 0:00:05.052819
+DEBUG:ibis.dask.trace:     execute_selection_dataframe Selection 0:00:05.054894
+DEBUG:ibis.dask.trace:   execute_until_in_scope Selection 0:00:05.055662
+DEBUG:ibis.dask.trace: main_execute Selection 0:00:05.056556
+"""
+
+_logger = logging.getLogger('ibis.dask.trace')
+
+# A list of funcs that is traced
+_trace_funcs = set()
+_trace_root = "main_execute"
+_TRACE_CONFIG = 'dask.enable_trace'
+
+
+def enable():
+    """Enable tracing."""
+    set_option(_TRACE_CONFIG, True)
+    logging.getLogger('ibis.dask.trace').setLevel(logging.DEBUG)
+
+
+def _log_trace(func, start=None):
+    level = 0
+    current_frame = None
+
+    # Increase the current level for each traced function in the stackframe
+    # This way we can visualize the call stack.
+    for frame, _ in traceback.walk_stack(None):
+        current_frame = current_frame if current_frame is not None else frame
+        func_name = frame.f_code.co_name
+        if func_name in _trace_funcs:
+            level += 1
+
+    # We can assume we have 'args' because we only call _log_trace inside
+    # trace or TraceDispatcher.resgister
+    current_op = current_frame.f_locals['args'][0]
+
+    # If the first argument is a Expr, we print its op because it's more
+    # informative.
+    if isinstance(current_op, ir.Expr):
+        current_op = current_op.op()
+
+    _logger.debug(
+        f"{'  ' * level} {func.__name__} {type(current_op).__qualname__} "
+        f"{datetime.now() - start if start else ''}"
+    )
+
+
+def trace(func):
+    """ Return a function decorator that wraped the decorated function with
+    tracing.
+    """
+    _trace_funcs.add(func.__name__)
+
```

Comment:
is this needed at all?

we can simply move the dispatcher / trace to a common module (maybe worthile doing that first).

you should be able to put in ibis/common (with the assocaited tests).

Reply:
This currently exists separately so that it has its own logger (`ibis.dask.trace`) and `_TRACE_CONFIG`  (`'dask.enable.trace`). 

I agree that this should get consolidated eventually, don't have a strong view on what we do with it now. Right now dask.core is importing all the pandas machinery anyway, so I'm happy to get rid of trace.py/dispatcher.py for now.

Reply:
yeah ok to do later

Reply:
https://github.com/ibis-project/ibis/issues/2551